cli/ops/websocket.rs

```diff
@@ -0,0 +1,221 @@
+      Some(header) => header.to_str().unwrap(),
+      None => "",
+    };
+    let extensions = response
+      .headers()
+      .get_all("Sec-WebSocket-Extensions")
+      .iter()
+      .map(|header| header.to_str().unwrap())
+      .collect::<String>();
+    Ok(json!({
+    "success": true,
+    "rid": rid,
+    "protocol": protocol,
+    "extensions": extensions
+    }))
+  };
+  Ok(JsonOp::Async(future.boxed_local()))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SendArgs {
+  rid: u32,
+  text: Option<String>,
+}
+
+pub fn op_ws_send(
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: SendArgs = serde_json::from_value(args)?;
+  let mut resource_table = isolate_state.resource_table.borrow_mut();
+  let stream = resource_table
+    .get_mut::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+    .ok_or_else(OpError::bad_resource_id)?;
+  block_on(stream.send(match args.text {
+    Some(text) => Message::Text(text),
+    None => Message::Binary(zero_copy[0].to_owned().to_vec()),
+  }))
+  .unwrap();
+  Ok(JsonOp::Sync(json!({})))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CloseArgs {
+  rid: u32,
+  code: Option<u16>,
+  reason: Option<String>,
+}
+
+pub fn op_ws_close(
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: CloseArgs = serde_json::from_value(args)?;
+  let resource_table = isolate_state.resource_table.clone();
+  let future = async move {
+    let mut stream = {
+      let mut resource_table = resource_table.borrow_mut();
+      resource_table
+        .remove::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+        .ok_or_else(OpError::bad_resource_id)?
+    };
+    (*stream)
+      .close(Some(CloseFrame {
+        code: CloseCode::from(args.code.unwrap_or(1005)),
+        reason: match args.reason {
+          Some(reason) => Cow::from(reason),
+          None => Default::default(),
+        },
+      }))
+      .await
+      .unwrap();
+
+    Ok(json!({}))
+  };
+
+  Ok(JsonOp::Async(future.boxed_local()))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct NextEventArgs {
+  rid: u32,
+}
+
+pub fn op_ws_next_event(
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: NextEventArgs = serde_json::from_value(args)?;
+  let resource_table = isolate_state.resource_table.clone();
+  let future = async move {
+    let mut resource_table = resource_table.borrow_mut();
+    let stream = resource_table
+      .get_mut::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+      .ok_or_else(OpError::bad_resource_id)?;
+    let message = match stream.next().await.unwrap() {
```

Comment:
borrow mut errors most likely originate from here, @crowlKats try to rewrite this future as a `poll_fn` so resource table is borrow only for the duration of the poll

Reply:
Unfortunately `WebSocketStream` (https://docs.rs/tokio-tungstenite/0.11.0/tokio_tungstenite/struct.WebSocketStream.html) doesn't expose poll versions of its methods. Not sure how to work around this.

Reply:
It does, it implements the `futures::Stream<Message>` trait, so it has a `poll_next` method: https://docs.rs/futures-core/0.3.5/futures_core/stream/trait.Stream.html

Reply:
It implements `Stream` trait so `poll_next` should be available

Reply:
I guess that works for this case, but what about [the close method](https://docs.rs/tokio-tungstenite/0.11.0/tokio_tungstenite/struct.WebSocketStream.html#method.close)?

Reply:
Doesn't matter for `close` as it's removed from resource table before calling and awaiting the method, hence we don't hold the borrow on resource table.